[BE] Team7-bat, yeon todo-list 구현

kimnayeon0108 opened the pull request:
안녕하세요! team7의 bat와 yeon입니다.

[배포 링크](http://ec2-3-36-241-44.ap-northeast-2.compute.amazonaws.com:8080/history)

[API 첨부 링크](https://github.com/kjk402/todo-list/wiki/API-url,-data-%EC%A0%95%EB%A6%AC) 

card의 생성, 삭제, 이동, 수정까지 구현하고, 이러한 action에 따른 history도 구현했습니다.
drag and drop 구현시 두 flag의 사이 값을 구하는 로직으로 구현했습니다.
card 값을 flag 순으로 정렬하는 쿼리를 작성해서 컬럼별로 card 리스트를 반환하도록 했습니다.
우선적으로 기능 구현을 진행했고, 예외처리는 추후 진행 예정입니다.

감사합니다.

kjk402 committed:
spring initialize

kjk402 committed:
README.MD 수정

kjk402 committed:
ios 파일 추가

kimnayeon0108 committed:
feat: Add configuration for connection with mysql

kimnayeon0108 committed:
feat: Add Card entity

kimnayeon0108 committed:
feat: Add schema.sql to add table in db

kimnayeon0108 committed:
feat: Add ApiCardController

kimnayeon0108 committed:
feat: Add card.sql

kimnayeon0108 committed:
test: Add test code to make sure saving and finding function works

kimnayeon0108 committed:
feat: Add configuration to connect with MySQL

kimnayeon0108 committed:
feat: Add Card entity and status enum

kimnayeon0108 committed:
feat: Create CardController, CardRepository, CardService

kimnayeon0108 committed:
feat: Add post type request controller

kimnayeon0108 committed:
feat: Delete test code

kimnayeon0108 committed:
feat: Delete @Sql annotation and add configuration to execute schema.sql

kimnayeon0108 committed:
Update README.md

kimnayeon0108 committed:
Update README.md

kimnayeon0108 committed:
Update README.md

kimnayeon0108 committed:
Update README.md

kimnayeon0108 committed:
Update README.md

kimnayeon0108 committed:
Update README.md

kimnayeon0108 committed:
Update README.md

kimnayeon0108 committed:
Update README.md

ITzombietux committed:
Update README.md

ITzombietux committed:
Update README.md

ITzombietux committed:
Update README.md

ITzombietux committed:
Update README.md

kimnayeon0108 committed:
feat: Add delete, move, update function
Add History class

kimnayeon0108 committed:
Resolve conflict to merge

kimnayeon0108 committed:
Merge branch 'dev' into pair/yeon

kimnayeon0108 committed:
Merge pull request #19 from kjk402/pair/yeon

Merge pair/yeon

kimnayeon0108 committed:
Delete ApiCardController.java

kimnayeon0108 committed:
test: Delete test code to deploy

kimnayeon0108 committed:
feat: Delete schema.sql file

kimnayeon0108 committed:
feat: Change logging level

kimnayeon0108 committed:
Change .gitignore file

kimnayeon0108 committed:
Change .gitignore file

kimnayeon0108 committed:
Change .gitignore file

kimnayeon0108 committed:
fixed untracked files

kimnayeon0108 committed:
Add .gitignore

kimnayeon0108 committed:
Change gitignore

kimnayeon0108 committed:
fixed untracked files

kimnayeon0108 committed:
feat: Add flag and view cards order by flag function

kimnayeon0108 committed:
feat: Add Table annotation

ksundong commented on README.md:
```diff
@@ -1,2 +1,71 @@
+todo-list 7팀
+
+### 개발 환경
+![xcode](https://user-images.githubusercontent.com/59398492/114123129-a7eefc00-992c-11eb-8819-67a13cd0f921.jpg)
+![intellij](https://user-images.githubusercontent.com/59398492/114123155-b76e4500-992c-11eb-9fab-bf298e70b01b.jpg)
+![mysql](https://user-images.githubusercontent.com/59398492/114123158-b89f7200-992c-11eb-9392-e8823d075aa7.png)
+![AWS](https://user-images.githubusercontent.com/59398492/114123162-b9380880-992c-11eb-9446-25729659d477.png)
```
오호 이미지 넣은건가요 ㅎㅎ

ksundong commented on README.md:
```diff
@@ -1,2 +1,71 @@
+
+[Wiki 링크](https://github.com/kjk402/todo-list/wiki/API-url,-data-%EC%A0%95%EB%A6%AC)
+
+### 소통창구
+
+정규 과정 시간에는 줌의 소회의실을 이용한다.
+
+iOS, 백엔드 클래스별로 나눠서 활동할 때와 정규 과정 시간 이후(6시 이후)에는 discord를 활용하여 활동사항을 공유한다.
+
+![https://user-images.githubusercontent.com/59398492/114122857-1da69800-992c-11eb-8f4b-b953b9b0c388.png](https://user-images.githubusercontent.com/59398492/114122857-1da69800-992c-11eb-8f4b-b953b9b0c388.png)
+
```
discord 사용 좋네요.

ksundong commented on README.md:
```diff
@@ -1,2 +1,71 @@
+- Network 구조 설계
+- URL을 구성하는 EndPoint 구조체 생성
+- Card를 가져오는 GET 요청
+- POST 요청 진행 중
```
이런 공통으로 쓰이는 부분은 BE 브랜치가 아니라 다른 브랜치에서 작업해주시는게 PR Merge할 때, 안전할 것 같긴해요!

ksundong commented on backend/.gitignore:
```diff
@@ -0,0 +1,39 @@
+/.nb-gradle/
+
+### VS Code ###
+.vscode/
```
기본으로 제공해주는 것 이외에 추가하시고 싶으시면 gitignore.io에 미리 정의된 항목들을 사용해보셔도 좋을 것 같아요.

ksundong commented on backend/build.gradle:
```diff
@@ -0,0 +1,28 @@
+}
+
+dependencies {
+	implementation group: 'org.springframework.boot', name: 'spring-boot-starter-data-jdbc', version: '2.2.5.RELEASE'
```
직접 추가해주신건가요? 버전이 혼자 2.2.5네요.
아래의 의존성들과 형식을 맞춰주시면 좋을 것 같아요.

kjk402 replied:
build.gradle 수정하였습니다.

ksundong commented on backend/src/main/java/com/codesquad/todolist/controller/CardController.java:
```diff
@@ -0,0 +1,55 @@
+import java.util.List;
+
+@RestController
+@RequestMapping("/card")
```
```suggestion
@RequestMapping("/cards")
```

오 card가 메인인 구조인가요~

kimnayeon0108 replied:
@ksundong url 부분은 iOS분들에게 이미 API를 제공한 상태라 변경이 어려울거 같습니다ㅠㅠ 다음부터 url에 더 신경쓰도록 하겠습니다!

ksundong commented on backend/src/main/java/com/codesquad/todolist/controller/CardController.java:
```diff
@@ -0,0 +1,55 @@
+        this.cardService = cardService;
+    }
+
+    @GetMapping("/{columnId}")
```
`/cards/{columnId}`인데 이 구조가 괜찮은걸까요?
`/cards/{cardId}`와 구분이 되지 않을 것 같아요.

kimnayeon0108 replied:
@ksundong columId 별로  즉, todo, doing, done 별로 카드를 보여주기 때문에 이런식으로 url을 작성하게 되었습니다. cardId와 헷갈릴수도 있는 문제가 생긴다면, url에 뭔가 더 추가해주는 구조가 더 좋을까요?

ksundong replied:
네 맞습니다. column에 card가 종속적이라면, url도 그렇게 바뀌는게 좋겠죠~

ksundong commented on backend/src/main/java/com/codesquad/todolist/controller/CardController.java:
```diff
@@ -0,0 +1,55 @@
+    }
+
+    @GetMapping("/{columnId}")
+    public ResponseEntity view(@PathVariable Long columnId) {
```
`ResponseEntity`를 직접 반환하는 형태는 지양해주세요.

kimnayeon0108 replied:
@ksundong response body 형태로 반환하도록 코드 수정 완료했습니다! responseBody 라는 메소드를 추가했습니다!

ksundong replied:
음..? DTO 를 반환하는 형태를 고민해주세요!

ksundong commented on backend/src/main/java/com/codesquad/todolist/controller/CardController.java:
```diff
@@ -0,0 +1,55 @@
+
+    @GetMapping("/{columnId}")
+    public ResponseEntity view(@PathVariable Long columnId) {
+        List<Card> cards = cardService.viewCardByColumnId(columnId);
```
`List`를 직접 반환하면 생길수 있는 문제는 List 이외의 정보를 담아야 하는 경우 breaking change가 발생합니다.

예를 들어...
```json
[
  {},
  {},
]
```

이런 구조로 개발되어 있는데, 특정 요소가 추가되어야 한다면 전체적으로 구조가 바뀌어야 합니다.

대신 이런구조를 고민해봐주세요.

```json
{
  "cards": []
}
```

이런 구조가 된다면, 유연하게 대응할 수 있게됩니다.

kimnayeon0108 replied:
@ksundong response body를 반환하도록 위와같이 코드 변경하니 말씀하신대로 구조가 변경되었습니다!

ksundong commented on backend/src/main/java/com/codesquad/todolist/controller/CardController.java:
```diff
@@ -0,0 +1,55 @@
+
+
+    @PostMapping("/{columnId}")
+    public ResponseEntity create(@PathVariable Long columnId, @RequestParam(value = "title") String title,
+                                 @RequestParam(value = "contents") String contents) {
```
title, contents로 받기보다는 DTO로 받는 방향을 고민해주세요.

kimnayeon0108 replied:
@ksundong 혹시 DTO는 다음번에 적용해봐도 괜찮을까요.....?ㅠㅠㅠ 시간이 부족할 듯 합니다ㅠㅠ 다음에 꼭 적용해보겠습니다!

ksundong replied:
네~

ksundong commented on backend/src/main/java/com/codesquad/todolist/controller/CardController.java:
```diff
@@ -0,0 +1,55 @@
+        return new ResponseEntity(card, HttpStatus.OK);
+    }
+
+    @DeleteMapping("/{id}")
+    public ResponseEntity delete(@PathVariable Long id) {
+        cardService.delete(id);
+        return new ResponseEntity(HttpStatus.OK);
+    }
```
delete는 굳이 결과를 주지 않아도 괜찮습니다.

ksundong replied:
204 응답 코드에 대해서 한 번 학습해보세요~

kimnayeon0108 replied:
@ksundong void로 메소드 반환타입 변경했습니다! 204 응답코드에 대해 학습해보겠습니다~🙌

ksundong commented on backend/src/main/java/com/codesquad/todolist/controller/CardController.java:
```diff
@@ -0,0 +1,55 @@
+        return new ResponseEntity(HttpStatus.OK);
+    }
+
+    @PutMapping("/{id}/move/{columnId}/{index}")
```
path 말고도 param으로 받는 방법도 있습니다.

ksundong commented on backend/src/main/java/com/codesquad/todolist/controller/HistoryController.java:
```diff
@@ -0,0 +1,22 @@
+        this.historyRepository = historyRepository;
+    }
+
+    @GetMapping("/history")
```
RequestMapping으로 history는 꺼내줘도 괜찮지 않을까요?

kimnayeon0108 replied:
@ksundong 네네 변경완료했습니다!

ksundong commented on backend/src/main/java/com/codesquad/todolist/controller/HistoryController.java:
```diff
@@ -0,0 +1,22 @@
+
+    public HistoryController(HistoryRepository historyRepository) {
+        this.historyRepository = historyRepository;
+    }
```
별도의 Service 의존성이 없는건가요? 가급적이면 Controller에서 Repository를 직접 사용하지는 않도록 해주세요.

kimnayeon0108 replied:
@ksundong HistoryService를 생성해서 컨트롤러에서 서비스를 의존하도록 변경했습니다!

ksundong commented on backend/src/main/java/com/codesquad/todolist/domain/Action.java:
```diff
@@ -0,0 +1,5 @@
+package com.codesquad.todolist.domain;
+
+public enum Action {
+    ADD, UPDATE, REMOVE, MOVE
```
```suggestion
    ADD,
    UPDATE,
    REMOVE,
    MOVE
```

Enum 사용 좋습니다.

ksundong commented on backend/src/main/java/com/codesquad/todolist/domain/Card.java:
```diff
@@ -0,0 +1,61 @@
+    private String contents;
+    private Long columnId;
+    private LocalDateTime createdTime;
+    private Double flag;
```
어떤 flag기에 double형인가요?

ksundong commented on backend/src/main/java/com/codesquad/todolist/domain/Card.java:
```diff
@@ -0,0 +1,61 @@
+
+import java.time.LocalDateTime;
+
+@Table("CARD")
```
일부러 테이블 명을 지정해주신건가요?

kjk402 replied:
네 맞습니다. 로컬개발 환경의 mysql과 서버에 올라간 mysql  의 대소문자 구분하는 lower_case_table 설정이 달라서 발생한 에러 수정할 때 대문자로 테이블명 지정해주었습니다.

ksundong commented on backend/src/main/java/com/codesquad/todolist/domain/Card.java:
```diff
@@ -0,0 +1,61 @@
+@Table("CARD")
+public class Card {
+
+    @Id
+    private Long id;
```
따로 GeneratedValue를 사용해주지 않은 이유가 있나요?

kjk402 replied:
database 테이블 설계 시 id에 Auto_Increment 속성을 추가하여 GeneratedValue  대신 사용해보았습니다.

ksundong replied:
스키마 파일을 보지는 못한 것 같은데 버전관리가 안되고 있는건가요?

kimnayeon0108 replied:
@ksundong 스키마 파일을 추가하긴 했는데, 버전관리가 어떤 의미로 말씀하시는건지 잘 모르겠습니다. 저희가 작성한 schema.sql이 올바르게 작성된건지 잘 모르겠습니다...😰

ksundong replied:
schema.sql 파일이 git으로 버전관리 되고있는 건가요? 파일을 보지 못한 것 같아서요!

kimnayeon0108 replied:
@ksundong 이전에는 MySQL에서 직접 테이블을 생성해주었고, schema파일은 이번에 새로 생성했습니다!

ksundong commented on backend/src/main/java/com/codesquad/todolist/domain/Card.java:
```diff
@@ -0,0 +1,61 @@
+    private String title;
+    private String contents;
+    private Long columnId;
+    private LocalDateTime createdTime;
```
```suggestion
    private LocalDateTime createdDateTime;
```

```suggestion
    private LocalDateTime createdAt;
```

ksundong commented on backend/src/main/java/com/codesquad/todolist/domain/Card.java:
```diff
@@ -0,0 +1,61 @@
+        return flag;
+    }
+
+    public void update(String title, String contents) {
+        this.title = title;
+        this.contents = contents;
+    }
+
+    public void move(Long toColumnId, Double flag) {
+        this.columnId = toColumnId;
+        this.flag = flag;
+    }
```
객체에게 요청하는 것 아주 좋습니다.

kjk402 replied:
ㅎㅎ 당근 감사합니다.~~

ksundong commented on backend/src/main/java/com/codesquad/todolist/domain/History.java:
```diff
@@ -0,0 +1,53 @@
+
+@Table("HISTORY")
+public class History {
+    private static final DateTimeFormatter FORMATTER_PATTERN = DateTimeFormatter.ofPattern("yyyy-MM-dd HH:mm");
```
History가 관리해야하는지는 조금 의문이네요.
FORMATTER_PATTERN 보다는 더 좋은 이름을 고민해봅시다.

ksundong commented on backend/src/main/java/com/codesquad/todolist/domain/History.java:
```diff
@@ -0,0 +1,53 @@
+    @Id
+    private Long id;
+
+    private String action;
```
Enum을 직접 사용하는 것은 고려해보지 않으셨나요?

kjk402 replied:
Enum 직접 사용하는 것으로 교체하였습니다.~

ksundong commented on backend/src/main/java/com/codesquad/todolist/domain/History.java:
```diff
@@ -0,0 +1,53 @@
+        return toColumnId;
+    }
+
+    public String getCreatedTime() {
+        return createdTime.format(FORMATTER_PATTERN);
+    }
```
getter를 바꾸는 대신에 다른 방법을 생각해봅시다.
Jackson을 이용하면 좋을 것 같네요.

kimnayeon0108 replied:
@ksundong 검색해보니 `@JsonFormat`이라는 어노테이션을 추가해서 getter로 바꾸는 형식을 대신했습니다.

ksundong commented on backend/src/main/java/com/codesquad/todolist/repository/CardRepository.java:
```diff
@@ -0,0 +1,30 @@
+
+public interface CardRepository extends CrudRepository<Card, Long> {
+
+    @Query("SELECT * FROM CARD WHERE CARD.COLUMN_ID=1 ORDER BY FLAG ASC")
```
`*`는 사용하지 말아주세요.

ksundong replied:
지금 보니까 FLAG는 Order 용으로 쓰이나보군요.

kjk402 replied:
네네 flag가 우선순위를 뜻합니다.
'*' 제거해주었습니다.

ksundong commented on backend/src/main/java/com/codesquad/todolist/repository/CardRepository.java:
```diff
@@ -0,0 +1,30 @@
+    @Query("SELECT * FROM CARD WHERE CARD.COLUMN_ID=3 ORDER BY FLAG ASC")
+    List<Card> findDoneCards();
+
+    @Query("SELECT FLAG FROM CARD WHERE CARD.COLUMN_ID=1 ORDER BY FLAG ASC")
+    List<Double> findTodoFlags();
+
+    @Query("SELECT FLAG FROM CARD WHERE CARD.COLUMN_ID=2 ORDER BY FLAG ASC")
+    List<Double> findDoingFlags();
+
+    @Query("SELECT FLAG FROM CARD WHERE CARD.COLUMN_ID=3 ORDER BY FLAG ASC")
+    List<Double> findDoneFlags();
```
어떤 용도의 메서드인지 궁금하네요.

ksundong commented on backend/src/main/java/com/codesquad/todolist/service/CardService.java:
```diff
@@ -0,0 +1,126 @@
+@Service
+public class CardService {
+
+    private final CardRepository cardRepository;
+    private final HistoryRepository historyRepository;
```
가급적이면 1 Service 1 Repository를 목표로 해주세요.
이것이 정답은 아니지만, 제가 경험해봤을 땐, 로직 분리에 도움이 되었습니다.

kjk402 replied:
네 historyService 생성하여 분리해주었습니다.

ksundong commented on backend/src/main/java/com/codesquad/todolist/service/CardService.java:
```diff
@@ -0,0 +1,126 @@
+        if (columnId == 3) {
+            return cardRepository.findDoneCards();
+        }
+        return null;
```
`null`을 리턴하지 말고 `new ArrayList<>();` 와 같은식으로 빈 컬렉션을 리턴해주세요.
앞으로 어떤 코드든 `null`을 리턴하는 일은 없다고 생각해주세요!

ksundong commented on backend/src/main/java/com/codesquad/todolist/service/CardService.java:
```diff
@@ -0,0 +1,126 @@
+    }
+
+    public List<Card> viewCardByColumnId(Long columnId) {
+        if (columnId == 1) {
+            return cardRepository.findTodoCards();
+        }
+        if (columnId == 2) {
+            return cardRepository.findDoingCards();
+        }
+        if (columnId == 3) {
+            return cardRepository.findDoneCards();
+        }
```
early return의 사용도 좋지만
`return cardRepository.findByColumnId(columnId);`
와 같은 식으로 풀면 어땠을까요?

kjk402 replied:
@Query 에 변수받아서 사용하는 것으로 수정하면 될 것 같습니다. 수정하겠습니다.

ksundong commented on backend/src/main/java/com/codesquad/todolist/service/CardService.java:
```diff
@@ -0,0 +1,126 @@
+        return null;
+    }
+
+    public Card create(Long columnId, String title, String contents) {
+        double flag = 10;
+        if (columnId == 1) {
+            List<Double> flags = new ArrayList<>(cardRepository.findTodoFlags());
+            if (flags.size() > 0) {
+                flag = flags.get(flags.size() - 1) + 10;
+            }
+        }
+        if (columnId == 2) {
+            List<Double> flags = new ArrayList<>(cardRepository.findDoingFlags());
+            if (flags.size() > 0) {
+                flag = flags.get(flags.size() - 1) + 10;
+            }
+        }
+        if (columnId == 3) {
+            List<Double> flags = new ArrayList<>(cardRepository.findDoneFlags());
+            if (flags.size() > 0) {
+                flag = flags.get(flags.size() - 1) + 10;
+            }
+        }
+        Card card = new Card(title, contents, columnId, flag);
+        cardRepository.save(card);
+        createHistory(Action.ADD.toString(), title, columnId, null);
+        return card;
+    }
```
이곳의 코드가 많이 복잡하죠?
설계에 문제가 있을 가능성이 매우 높아보이네요.

그리고, 만약 card가 생성되고, history를 생성하는 도중에 오류가 발생한다면 어떻게 될까요?
두 분 상의해서 답변해주세요.

ksundong replied:
if 문을 최소화시켜보시고, 중복되는 로직은 메서드로 추출해보세요!

kimnayeon0108 replied:
@ksundong transaction으로 card와 history의 작업을 하나로 묶어주면 괜찮을까요...?

ksundong replied:
네 후자는 트랜잭션을 사용해주면 될겁니다.
로직 개선은 더 고민해봐주세요~

ksundong commented on backend/src/main/java/com/codesquad/todolist/service/CardService.java:
```diff
@@ -0,0 +1,126 @@
+        }
+        Card card = new Card(title, contents, columnId, flag);
+        cardRepository.save(card);
+        createHistory(Action.ADD.toString(), title, columnId, null);
```
인자로 `null`을 넣어주는 행위 역시 지양해주세요.

kimnayeon0108 replied:
@ksundong 0으로 값을 넣어주는것으로 대체했습니다. 이부분도 앞으로 신경써야겠네요!

ksundong replied:
새로운 메서드를 정의해줘도 됩니다~

ksundong commented on backend/src/main/java/com/codesquad/todolist/service/CardService.java:
```diff
@@ -0,0 +1,126 @@
+        historyRepository.save(history);
+    }
+
+    private Card findCard(Long id) {
+        return cardRepository.findById(id).orElse(null);
+    }
```
이 코드 다시한 번 생각해봐주세요.
`Optional.orElse(null)`은 올바른 방식이 아니에요 ㅠㅠㅠ

kjk402 replied:
orElseThrow 사용하도록 코드 변경하였습니다.

ksundong commented on backend/src/main/java/com/codesquad/todolist/service/CardService.java:
```diff
@@ -0,0 +1,126 @@
+
+    public Card update(Long id, String title, String contents) {
+        Card card = findCard(id);
+        card.update(title, contents);
```
NPE가 발생할 수 있겠죠?

kjk402 replied:
NPE의 경우 ios앱에서 값을 입력하지 않을 경우 생성버튼을 비활성화 하는 방식으로 
진행협의하여 null값을 받을 가능성을 제거해서 따로 신경쓰지 않았습니다. ^^

그렇다 하더라도 백엔드에서도 null값 vaild가 필요할까요?

ksundong replied:
iOS랑 아무 상관없이 null을 리턴하는 findCard라는 메서드가 있는 한 NPE의 가능성이 존재합니다.
클라이언트는 너무 믿지 말고, 항상 모든 경우의 수를 염두에 두는 것을 추천합니다.

kjk402 replied:
넵 알겠습니다. nullCheck 메소드 생성하였습니다. ㅎㅎ

ksundong replied:
null이 발생하지 않도록 설계하는 것도 좋겠죠.

ksundong commented on backend/src/main/java/com/codesquad/todolist/service/CardService.java:
```diff
@@ -0,0 +1,126 @@
+        return card;
+    }
+
+    private void createHistory(String action, String title, Long fromColumnId, Long toColumnId) {
```
`History`에 정적 팩토리 메서드로 만드는 방식을 고려해봐주세요.

ksundong replied:
그리고 Enum을 그대로 받는 편이 좋겠죠?

kjk402 replied:
팩토리 메소드로 교체하였습니다. 
Enum 그대로 받게 하였습니다.

ksundong commented on backend/src/main/java/com/codesquad/todolist/service/CardService.java:
```diff
@@ -0,0 +1,126 @@
+        return cardRepository.findById(id).orElse(null);
+    }
+
+    private Double updateFlag(List<Double> flags, int index) {
+        double flag = 0;
+        if (flags.size() == 0) {
+            flag = 10.0;
+            return flag;
+        }
+        if (index == 1) {
+            flag = (0 + flags.get(0)) / 2;
+            return flag;
+        }
+        if (flags.size() < index) {
+            flag = flags.get(flags.size() - 1) + 10;
+            return flag;
+        }
+        flag = (flags.get(index - 2) + flags.get(index - 1)) / 2;
+        return flag;
+    }
```
굉장히 로직이 복잡한데요, 단순 순서를 정하는게 이렇게 복잡한 로직을 사용해야 하는가? 하는 의문이 드네요.

ksundong reviewed (approved):
안녕하세요! Bat & Yeon!! 리뷰어 Dion입니다. 첫번째 PR 보내신것을 축하드려요!
처음 협업을 시작한 만큼 이것저것 신경쓸 것도 많고, 여러가지 시행착오들을 많이 겪으셨을 것 같습니다.
예외 처리는 안되어있더라도, 예외를 미리 정의해주시는 것은 추천드리겠습니다.
그리고, 재사용 할 수 있는 코드들을 많이 만들어두시면 앞으로 금방금방 개발을 하실 수 있을 거에요.

- column 관련 로직이 많이 아쉬웠습니다. 어떻게 개선할 수 있을까요?
- flag 관련 로직이 복잡해서, 간소화 하면 어떨까 싶습니다.
- REST API 설계가 조금 아쉽습니다. url과 response에 대해서 코멘트를 남겨두었으니 참고해주세요. (현재 단계에서 적용하지 않으셔도 됩니다.)

미션 진행하시느라 고생하셨습니다. 다음 단계에서 보다 개선되었으면 좋겠네요~

ksundong commented:
API 문서도 앞으로 좀 더 발전시켜 나가봅시다! 작성하느라 고생하셨어요!

kimnayeon0108 committed:
fix: Simply finding cards and flags query and Add factory method of History and nullCheck method

kimnayeon0108 committed:
feat: Add schema.sql and properties to execute schema.sql

kimnayeon0108 committed:
fix: Fix the way to format createdTime in domain and add responseBody

Add JasonFormat annotation to format createdTime.
Return response body instead of entity.

kimnayeon0108 committed:
fix: Add HistoryService

kimnayeon0108 committed:
fix: Remove useless code

ksundong commented on backend/src/main/java/com/codesquad/todolist/controller/HistoryController.java:
```diff
@@ -0,0 +1,32 @@
+        return ResponseEntity.ok(responseBody("history", historyService.findHistory()));
+    }
+
+    private Map<String, Object> responseBody(String rootName, Object body) {
+        Map<String, Object> responseBody = new HashMap<>();
+        responseBody.put(rootName, body);
+        return responseBody;
+    }
```
Map을 쓰진 말아주세요~

ksundong commented on backend/src/main/java/com/codesquad/todolist/domain/Card.java:
```diff
@@ -0,0 +1,65 @@
+
+    private String title;
+    private String contents;
+    private Long columnId;
```
이렇게 Long으로 쓰고 있는 부분을 DM으로 말씀드린거였어요 ㅎㅎ

kimnayeon0108 replied:
@ksundong 아하 이부분이였군요!🧐

ksundong commented on backend/src/main/java/com/codesquad/todolist/domain/History.java:
```diff
@@ -0,0 +1,58 @@
+    private Long fromColumnId;
+    private Long toColumnId;
+
+    @JsonFormat(shape = JsonFormat.Shape.STRING, pattern = "yyyy-MM-dd HH:mm", timezone = "Asia/Seoul")
```
이렇게 Annotation을 써줘도 좋고, 나중에 여러 개를 써야할 경우 중복 제거도 고민해주세요~

ksundong commented on backend/src/main/java/com/codesquad/todolist/repository/CardRepository.java:
```diff
@@ -0,0 +1,18 @@
+
+public interface CardRepository extends CrudRepository<Card, Long> {
+
+    @Query("SELECT ID, TITLE, CONTENTS, COLUMN_ID, CREATED_TIME, FLAG FROM CARD WHERE CARD.COLUMN_ID LIKE CONCAT('%',:columnId,'%') ORDER BY FLAG ASC")
```
과연 LIKE 쿼리를 써야했을까요?
그리고 이건 좀 어려운 문젠데, '%검색어%' 는 문제가 있습니다.
@honux77 이 부분 리뷰 해도 되나요?

ksundong replied:
일단은 =으로 충분히 검색 가능하니 그 부분을 고려해주세요.

ksundong commented on backend/src/main/java/com/codesquad/todolist/service/CardService.java:
```diff
@@ -0,0 +1,96 @@
+    }
+
+    public Card create(Long columnId, String title, String contents) {
+        double flag = 10;
```
10은 매직넘버죠? 추출해봅시다.

ksundong commented on backend/src/main/java/com/codesquad/todolist/service/CardService.java:
```diff
@@ -0,0 +1,96 @@
+
+    public Card create(Long columnId, String title, String contents) {
+        double flag = 10;
+        List<Double> flags = new ArrayList<>(cardRepository.findFlagsByColumnId(columnId));
+        if (flags.size() > 0) {
+            flag = flags.get(flags.size() - 1) + 10;
+        }
```
로직이 있는 컬렉션이라면 일급 컬렉션의 사용을 검토해주세요.

ksundong commented on backend/src/main/java/com/codesquad/todolist/service/CardService.java:
```diff
@@ -0,0 +1,96 @@
+        return cardRepository.findCardsByColumnId(columnId);
+    }
+
+    public Card create(Long columnId, String title, String contents) {
+        double flag = 10;
+        List<Double> flags = new ArrayList<>(cardRepository.findFlagsByColumnId(columnId));
+        if (flags.size() > 0) {
+            flag = flags.get(flags.size() - 1) + 10;
+        }
+        Card card = new Card(title, contents, columnId, flag);
+        cardRepository.save(card);
+        historyService.createHistory(Action.ADD, title, columnId, 0L);
+        return card;
+    }
```
빈 라인을 의미단위로 넣어서 쪼개봅시다.

ksundong commented on backend/src/main/java/com/codesquad/todolist/service/CardService.java:
```diff
@@ -0,0 +1,96 @@
+        historyService.createHistory(Action.REMOVE, card.getTitle(), card.getColumnId(), 0L);
+    }
+
+    public Card move(Long id, Long toColumnId, int index) {
+        Card card = findCard(id);
+        Long fromColumnId = card.getColumnId();
+        double flag = 0;
+        List<Double> flags = new ArrayList<>(cardRepository.findFlagsByColumnId(toColumnId));
+        flag = updateFlag(flags, index);
+        card.move(toColumnId, flag);
+        cardRepository.save(card);
+        historyService.createHistory(Action.MOVE, card.getTitle(), fromColumnId, toColumnId);
+        return card;
+    }
```
여기도 한 호흡으로 읽기엔 많이 깁니다.
빈 라인을 두어서 쪼개봅시다.

ksundong commented on backend/src/main/java/com/codesquad/todolist/service/CardService.java:
```diff
@@ -0,0 +1,96 @@
+        return card;
+    }
+
+    public void nullCheck(String title, String contents) {
+        if (title == null || title.trim().length() == 0) {
+            throw new NullPointerException("title can't be null");
+        }
+        if (contents == null || contents.trim().length() == 0) {
+            throw new NullPointerException("contents can't be null");
+        }
+    }
```
nullCheck가 CardService에 있어야 할 메서드일까요?
고민해봅시다.

ksundong commented on backend/src/main/java/com/codesquad/todolist/service/CardService.java:
```diff
@@ -0,0 +1,96 @@
+    }
+
+    private Card findCard(Long id) {
+        return cardRepository.findById(id).orElseThrow(
+                () -> new IllegalArgumentException());
```
```suggestion
        return cardRepository.findById(id).orElseThrow(IllegalArgumentException::new);
```

ksundong commented on backend/src/main/java/com/codesquad/todolist/service/HistoryService.java:
```diff
@@ -0,0 +1,28 @@
+    }
+
+    public List<History> findHistory() {
+        List<History> history = (List<History>) historyRepository.findAll();
```
여기서 캐스팅해주는 것 말고 다른 방법을 생각해봅시다.

ksundong commented on backend/src/main/resources/schema.sql:
```diff
@@ -0,0 +1,21 @@
+drop table if exists CARD;
+
+create table CARD (
+                      id bigint auto_increment primary key,
+                      title varchar(50),
+                      contents varchar(500),
+                      column_id int,
+                      created_time timestamp,
+                      flag double
+);
+
+drop table if exists HISTORY;
+
+create table HISTORY (
+                         id bigint auto_increment primary key,
+                         action varchar(10),
+                         title varchar(50),
+                         from_column_id bigint,
+                         to_column_id bigint,
+                         created_time timestamp
+);
```
확인했습니다~

ksundong reviewed (approved):
피드백 반영해주시느라 고생하셨습니다~ 화이팅!!

kimnayeon0108 commented:
@ksundong 머지 감사합니다!🙇‍♀️ 추가적인 피드백 주신건 bat랑 상의 후 수정해보겠습니다!

kjk402 commented:
@ksundong 리뷰 감사드립니다. 추가로 남겨주신 코멘트도 확인하겠습니다.

감사합니다~
